ROS2: converted to ros2

klintan opened the pull request:
Started ROS2 conversion. Still some minor issues and some things that should probably be made more "ROS2"-ish.

Still some linking issues in the cmakelists because it dynamically links libopenvslam, but image is missing when trying to run. Will try to fix asap.

Thought I'd put this out here in the meantime if someone wants to contribute or take a look and give some feedback on the work done so far.

EDIT: Forgot, if you guys could create a ROS2 branch, that would be awesome, and then I'll change this PR to be to the ROS2 branch instead.

MikiyaShibuya commented:
Thank you very much! I will review your PR soon.
Your ROS2 implementation is very nice, but I think ROS1 is also necessary for some users.
So, could you make some changes to keep both packages of ROS1 and ROS2 as bellow?
* /ros/1/[sources for ROS1]
* /ros/2/[sources for ROS2]

klintan commented:
> Thank you very much! I will review your PR soon.
> Your ROS2 implementation is very nice, but I think ROS1 is also necessary for some users.
> So, could you make some changes to keep both packages of ROS1 and ROS2 as bellow?
> 
> * /ros/1/[sources for ROS1]
> * /ros/2/[sources for ROS2]

Of course, if that's the structure you prefer. Another alternative, as mentioned, is to create a ROS2 branch and keep master in ROS1. 

But again I don't really have any preference, so I can def refactor in to that structure.

klintan commented:
Fixed the project structure, not sure why clang-format fails and I can't see the specific error (auth issue).

Let me know if you find the linking issue in the CMakeLists.txt file.

MikiyaShibuya commented:
Thank you!
That error was caused by an older version of wercker.yaml.
Please merge up-to-date master branch, then the error will be fixed.

klintan commented:
> Thank you!
> That error was caused by an older version of wercker.yaml.
> Please merge up-to-date master branch, then the error will be fixed.

Ah ok , I'll rebase this PR on top of master.

MikiyaShibuya commented:
I'm sorry for making you wait for a long time.
I'm checking your change now, and it will take more time.
Please wait for a couple of days.

MikiyaShibuya commented:
I'm working to determine if your implementations run properly.
Now, the ROS 1 package is working well on the procedure written in the OpenVSLAM documentation.
But in the ROS 2 package, openvslam topic couldn't receive any image inputs at all(all ros-topics don't dump any errors when running).
I used ROS2-dashing and colcon build on Ubuntu18.
Could you tell me how to build and run, and what OS you use?

klintan commented:
> I'm working to determine if your implementations run properly.
> Now, the ROS 1 package is working well on the procedure written in the OpenVSLAM documentation.
> But in the ROS 2 package, openvslam topic couldn't receive any image inputs at all(all ros-topics don't dump any errors when running).
> I used ROS2-dashing and colcon build on Ubuntu18.
> Could you tell me how to build and run, and what OS you use?

I run on mac. Hm ok strange, I'll take an extra look and see if I have the same issue.

I had to update a dynamic link to the openvslam library though.

MikiyaShibuya commented:
Thanks.
Also,  I want to know ROS2 distro you use, build framework(really ament_cmake?) and ROS node launching commands(like "$ ros2 run ~~").

klintan commented:
> Thanks.
> Also, I want to know ROS2 distro you use, build framework(really ament_cmake?) and ROS node launching commands(like "$ ros2 run ~~").

I'm using the latest dashing-diademata patch 3 on Mac 10.14 (Mojave). Yes ament is used for all Ros2, using colcon. 

The run command I run is:
```
ros2 run openvslam run_slam -v /openvslam/orb_vocab/orb_vocab.dbow2 -c /openvslam/usb_config.yaml
```

However had an issue that I have to investigate seen below:

```
Andreass-MacBook-Pro:openvslam andreasklintberg$ ros2 run openvslam run_slam -v /Users/andreasklintberg/personal/openvslam/orb_vocab/orb_vocab.dbow2 -c /Users/andreasklintberg/personal/openvslam/usb_config.yaml
[2019-11-30 14:06:42.763] [I] config file loaded: /Users/andreasklintberg/personal/openvslam/usb_config.yaml
  ___               __   _____ _      _   __  __
 / _ \ _ __  ___ _ _\ \ / / __| |    /_\ |  \/  |
| (_) | '_ \/ -_) ' \\ V /\__ \ |__ / _ \| |\/| |
 \___/| .__/\___|_||_|\_/ |___/____/_/ \_\_|  |_|
      |_|

Copyright (C) 2019,
National Institute of Advanced Industrial Science and Technology (AIST)
All rights reserved.

This is free software,
and you are welcome to redistribute it under certain conditions.
See the LICENSE file.

Camera Configuration:
- name: USB monocular
- setup: Monocular
- fps: 10
- cols: 640
- rows: 480
- color: RGB
- model: Perspective
  - fx: 4232.48
  - fy: 3709.79
  - cx: 437.474
  - cy: 107.442
  - k1: 0
  - k2: 0
  - p1: 0
  - p2: 0
  - k3: 0
  - min x: 0
  - max x: 640
  - min y: 0
  - max y: 480
ORB Configuration:
- number of keypoints: 1000
- scale factor: 1.2
- number of levels: 8
- initial fast threshold: 20
- minimum fast threshold: 7

[2019-11-30 14:06:42.764] [I] loading ORB vocabulary: /Users/andreasklintberg/personal/openvslam/orb_vocab/orb_vocab.dbow2
[2019-11-30 14:06:43.336] [I] startup SLAM system
[2019-11-30 14:06:43.336] [I] start global optimization module
[2019-11-30 14:06:43.336] [I] start mapping module
2019-11-30 14:06:43.403 run_slam[16206:1491189] *** Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: 'nextEventMatchingMask should only be called from the Main Thread!'
*** First throw call stack:
(
	0   CoreFoundation                      0x00007fff4fc20acd __exceptionPreprocess + 256
	1   libobjc.A.dylib                     0x00007fff7a326a17 objc_exception_throw + 48
	2   AppKit                              0x00007fff4d15ff6e -[NSApplication(NSEvent) _nextEventMatchingEventMask:untilDate:inMode:dequeue:] + 4180
	3   libpangolin.dylib                   0x0000000104c68cdd +[PangolinNSApplication run_step] + 109
	4   libpangolin.dylib                   0x0000000104c6784c _ZN8pangolin9OsxWindowC2ERKNSt3__112basic_stringIcNS1_11char_traitsIcEENS1_9allocatorIcEEEEiib + 252
	5   libpangolin.dylib                   0x0000000104c68737 _ZZN8pangolin24RegisterOsxWindowFactoryEvEN16OsxWindowFactory4OpenERKNS_3UriE + 1047
	6   libpangolin.dylib                   0x0000000104beabe0 _ZN8pangolin15FactoryRegistryINS_14VideoInterfaceEE4OpenERKNS_3UriE + 272
	7   libpangolin.dylib                   0x0000000104be9b3e _ZN8pangolin19CreateWindowAndBindENSt3__112basic_stringIcNS0_11char_traitsIcEENS0_9allocatorIcEEEEiiRKNS_6ParamsE + 1294
	8   libpangolin_viewer.dylib            0x0000000104b74fa3 _ZN15pangolin_viewer6viewer3runEv + 99
	9   run_slam                            0x000000010421c050 _ZNSt3__114__thread_proxyINS_5tupleIJNS_10unique_ptrINS_15__thread_structENS_14default_deleteIS3_EEEEZ13mono_trackingRKNS_10shared_ptrIN9openvslam6configEEERKNS_12basic_stringIcNS_11char_traitsIcEENS_9allocatorIcEEEESK_bSK_E3$_1EEEEEPvSN_ + 48
	10  libsystem_pthread.dylib             0x00007fff7bce92eb _pthread_body + 126
	11  libsystem_pthread.dylib             0x00007fff7bcec249 _pthread_start + 66
	12  libsystem_pthread.dylib             0x00007fff7bce840d thread_start + 13
)
libc++abi.dylib: terminating with uncaught exception of type NSException
```

klintan commented:
Fixed the publishers now, both image and video.

klintan commented:
> 2019-11-30 14:06:43.403 run_slam[16206:1491189] *** Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: 'nextEventMatchingMask should only be called from the Main Thread!'
*** First throw call stack:
(
	0   CoreFoundation                      0x00007fff4fc20acd __exceptionPreprocess + 256
	1   libobjc.A.dylib                     0x00007fff7a326a17 objc_exception_throw + 48
	2   AppKit                              0x00007fff4d15ff6e -[NSApplication(NSEvent) _nextEventMatchingEventMask:untilDate:inMode:dequeue:] + 4180
	3   libpangolin.dylib                   0x0000000104c68cdd +[PangolinNSApplication run_step] + 109
	4   libpangolin.dylib                   0x0000000104c6784c _ZN8pangolin9OsxWindowC2ERKNSt3__112basic_stringIcNS1_11char_traitsIcEENS1_9allocatorIcEEEEiib + 252
	5   libpangolin.dylib                   0x0000000104c68737 _ZZN8pangolin24RegisterOsxWindowFactoryEvEN16OsxWindowFactory4OpenERKNS_3UriE + 1047
	6   libpangolin.dylib                   0x0000000104beabe0 _ZN8pangolin15FactoryRegistryINS_14VideoInterfaceEE4OpenERKNS_3UriE + 272
	7   libpangolin.dylib                   0x0000000104be9b3e _ZN8pangolin19CreateWindowAndBindENSt3__112basic_stringIcNS0_11char_traitsIcEENS0_9allocatorIcEEEEiiRKNS_6ParamsE + 1294
	8   libpangolin_viewer.dylib            0x0000000104b74fa3 _ZN15pangolin_viewer6viewer3runEv + 99
	9   run_slam                            0x000000010421c050 _ZNSt3__114__thread_proxyINS_5tupleIJNS_10unique_ptrINS_15__thread_structENS_14default_deleteIS3_EEEEZ13mono_trackingRKNS_10shared_ptrIN9openvslam6configEEERKNS_12basic_stringIcNS_11char_traitsIcEENS_9allocatorIcEEEESK_bSK_E3$_1EEEEEPvSN_ + 48
	10  libsystem_pthread.dylib             0x00007fff7bce92eb _pthread_body + 126
	11  libsystem_pthread.dylib             0x00007fff7bcec249 _pthread_start + 66
	12  libsystem_pthread.dylib             0x00007fff7bce840d thread_start + 13
)
libc++abi.dylib: terminating with uncaught exception of type NSException****

Ok this issue is just that on MacOSX (recent versions at least) you can't run any UI stuff (Pangolin viewer) in a background thread, so I'll just have to put the Pangolin viewer in the main thread and the ros spin in a background thread. Hopefully that will fix it. Have you tested this in ROS1, is it working there ?

MikiyaShibuya commented:
Thank you so much!! Your comments and contributions are very helpful.
Now, I checked that run_slam and run_localization with pangolin-viewer are working properly.

Could you check the following points?

MikiyaShibuya commented on ros2/src/openvslam/src/run_localization.cc:
```diff
@@ -0,0 +1,206 @@
+            std::this_thread::sleep_for(std::chrono::microseconds(5000));
+        }
+        rclcpp::shutdown();
+        }
```
Please correct the indent.
CI error might be caused by it.

MikiyaShibuya commented on ros2/src/openvslam/src/run_localization.cc:
```diff
@@ -0,0 +1,206 @@
+        rclcpp::shutdown();
+        }
+#elif USE_SOCKET_PUBLISHER
+    std::thread thread([&]() {
```
When building with USE_SOCKET_PUBLISHER, compile error happened in that line like this.
` error: redeclaration of "std::thread thread" `
It would be helpful if you handle the case of using Socket-Publisher.

klintan commented:
fixed the issues you pointed out, but now for some reason the Pangolin library is not built in travis, although I have a really hard time seeing how these last changes could have affected that, so I'm leaning towards some kind of dependency versions issue. It builds fine locally.

klintan commented:
OK there was another issue which I fixed: when ros2 was sourced, the `yaml_cpp_vendor` is conflicting with `yaml-cpp` which results in the segmentation fault (similar to this https://github.com/jbeder/yaml-cpp/issues/299). I modified `DYLD_LIBRARY_PATH` env variable to not contain the  `yaml_cpp_vendor` and it works. Not sure what the final solution is, but I think it should be fine for now.

So this should be done now, as soon as Pangolin builds in Travis again, let me know if you have time to take a look at that, if not I'll try to dig in.

klintan commented:
@MikiyaShibuya any news on this ? would love to get this one rolling :) Let me know if I can do anything to get this going.

klintan commented:
@shinsumicco or @MikiyaShibuya I understand you are very busy, but just want to circle back on this PR. Would be awesome to get this into the openvslam lib.

shinsumicco commented:
@klintan Very sorry for making you wait for a long time.
I'll check the behavior of your pull-request on docker-based ROS2 environment this weekend.
Thank you for your patience and continued support.

klintan commented:
> @klintan Very sorry for making you wait for a long time.
> I'll check the behavior of your pull-request on docker-based ROS2 environment this weekend.
> Thank you for your patience and continued support.

No worries at all :) looking forward to the feedback.

shinsumicco commented:
@klintan Sorry for the late reply,
I'd built ROS2 packages according to:
your instruction: https://github.com/xdspacelab/openvslam/issues/247#issuecomment-589909116
@mirellameelo 's instruction: https://github.com/xdspacelab/openvslam/issues/247#issuecomment-591665991

But following error messages was appeared:
```
$ pwd
/home/ubuntu/Programs/openvslam/ros2
$ colcon build --symlink-install
Starting >>> cv_bridge
Starting >>> image_transport
Starting >>> camera_calibration_parsers
Starting >>> image_geometry
Finished <<< camera_calibration_parsers [12.7s]                                                         
Finished <<< image_geometry [14.6s]                                                                     
Finished <<< cv_bridge [22.6s]                                                                            
Starting >>> opencv_tests                                    
Starting >>> vision_opencv
Finished <<< image_transport [22.9s]                                                               
Starting >>> camera_info_manager
Starting >>> openvslam
Starting >>> publisher
Finished <<< opencv_tests [1.29s]                                                        
Finished <<< vision_opencv [2.50s]                                                                      
--- stderr: publisher                                                                                 
/usr/bin/ld: warning: libopencv_imgcodecs.so.3.2, needed by /opt/ros/dashing/lib/libcv_bridge.so, may conflict with libopencv_imgcodecs.so.3.4
/usr/bin/ld: warning: libopencv_core.so.3.2, needed by /opt/ros/dashing/lib/libcv_bridge.so, may conflict with libopencv_core.so.3.4
/usr/bin/ld: warning: libopencv_core.so.3.2, needed by /opt/ros/dashing/lib/libcv_bridge.so, may conflict with libopencv_core.so.3.4
---
Finished <<< publisher [2.49s]
Finished <<< camera_info_manager [5.34s]                                                                 
--- stderr: openvslam                                  
/usr/bin/ld: warning: libopencv_imgcodecs.so.3.2, needed by /opt/ros/dashing/lib/libcv_bridge.so, may conflict with libopencv_imgcodecs.so.3.4
/usr/bin/ld: warning: libopencv_core.so.3.2, needed by /opt/ros/dashing/lib/libcv_bridge.so, may conflict with libopencv_core.so.3.4
/usr/bin/ld: warning: libopencv_imgcodecs.so.3.2, needed by /opt/ros/dashing/lib/libcv_bridge.so, may conflict with libopencv_imgcodecs.so.3.4
/usr/bin/ld: warning: libopencv_core.so.3.2, needed by /opt/ros/dashing/lib/libcv_bridge.so, may conflict with libopencv_core.so.3.4
---
Finished <<< openvslam [14.9s]
                      
Summary: 9 packages finished [38.0s]
  2 packages had stderr output: openvslam publisher
```

What caused these errors?
It seems that OpenCVs which are installed in the system and ROS2 are conflicted each other.
Should I link the OpenCV binded with ROS2 to OpenVSLAM?

mirellameelo commented:
@shinsumicco hi! About the warnings, I had those too and I was able to run with the dataset with no problem. I intend to test today with a webcam and then add my step by step to others to try. 

About the errors: you are building again the image_common and vision_opencv packages, this is because the folders are inside your openvslam folder, right? "Colcon build" install packages recursively... Try to remove them to another path folder, build again, one by one, source both and only after that, build openvslam packages. As you can see in my comment, I used a different path folder for each other. 

But, if is really important to you keep in the same folder, you can use the following command:

```bh
colcon build --packages-select <package name>

#our case
colcon build --packages-select openvslam
```

shinsumicco commented:
@mirellameelo Thank you for the information!
The instructions you posted are very helpful to us.
Finally I succeeded to build the packages and launch ROS nodes.
@klintan Please let me review this pull-request. 
Thank you for your patience.

klintan commented:
I know things are a bit crazy now, I hope you are all doing all right! I was wondering if you have had the chance to take a look at these now? 
Thanks @mirellameelo for providing these instructions!

shinsumicco commented:
Very sorry to keep you waiting.
I've merged the PR after the following fixes:

- make the dependencies clear
- fix some coding styles

Our team member @MikiyaShibuya will prepare instructions for using ROS1/2 packages in the document.
I'd appreciate if you could confirm it after he finishes writing it.
Thank you for your continued support!

klintan commented:
> @klintan
> Very sorry to keep you waiting.
> I've merged the PR after the following fixes:
> 
> * make the dependencies clear
> * fix some coding styles
> 
> Our team member @MikiyaShibuya will prepare instructions for using ROS1/2 packages in the document.
> I'd appreciate if you could confirm it after he finishes writing it.
> Thank you for your continued support!

No worries at all, I'm just happy it got in there. Thank you for your patience on changes and back and forth as well.
